### PR TITLE
Ft/kms

### PR DESCRIPTION
**Pull request by ploki**
Key Management System (for server side encryption usage)

please, review the last commit, as others are part of different pull requests.

### CODE REVIEW

File: lib/kms/wrapper.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            }
+            return cb(err, cipherBundle);
+        });
+    }
+    /*
+     * cb(err, decipherBundle)
+     */
+    static createDecipherBundle(serverSideEncryption, offset,
+                                log, cb) {
+        const decipherBundle = {
```
Comment (ThibaultRiviere) [inline lib/kms/wrapper.js]: Maybe you can declare the variable after check the argument

**Comment by ThibaultRiviere** [comment]
:+1:

File: lib/kms/common.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
+import crypto from 'crypto';
+import { errors } from 'arsenal';
+
+export default class common {
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: capitalize class name "Common"

File: lib/kms/common.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                });
+        } else {
+            log.error('Unknown cryptographic scheme', { cryptoScheme });
+            cb(errors.InternalError);
+        }
+    }
+
+    /*
+     * cb(err, decipher: ReadWritable.stream)
+     */
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: Please add full jsdoc

File: lib/kms/file/backend.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import common from '../common';
+
+export const backend = {
+    /*
+     * Target implementation will be async in some aspects. let mimic it
```
Comment (LaurenSpiegel) [inline lib/kms/file/backend.js]: in some aspects?

let s/b let's

File: lib/kms/wrapper.js [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            return cb(null, masterKeyId);
+        });
+    }
+
+    /*
+     * cb(err, serverSideEncryptionInfo: object)masterKeyId: string)
+     */
+    static bucketLevelEncryption(bucketName, headers, log, cb) {
+        const sseAlgorithm = headers['x-amz-scal-server-side-encryption'];
+        const sseMasterKeyId =
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: do they need to send this?

File: lib/kms/common.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    static _IVSize() {
+        return 16;
+    }
+
+    /* block size is 128bit for AES-256-CTR */
+    static _aesBlockSize() {
+        return 16;
+    }
+
+    static createDataKey() {
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: Let's add a comment:
Creates data key to encrypt and decrypt the actual data (which data key is ciphered and deciphered by the appliance).  The encrypted data key is stored in the object's metadata. We also use this function to create the bucket key for the file and in memory implementations.

File: lib/kms/common.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            /* we are not storing hashed human password.
+             * It's a random key, so 1 iteration and
+             * a fixed salt is enough for our usecase.
+             * don't change these without bumping
+             * scheme number
+             */
+            const salt = new Buffer('ItsTasty');
+            const iterations = 1;
+            crypto.pbkdf2(
+                dataKey, salt, iterations,
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: Let's add a comment:
dataKey here is the unencrypted key (either from the appliance on a get or originally generated by kms in the case of a put)

File: lib/kms/common.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                this._keySize(), 'sha1', (err, derivedKey) => {
+                    if (err) {
+                        log.error('pbkdf2 function failed on key derivation',
+                                  { err });
+                        cb(errors.InternalError);
+                        return;
+                    }
+                    crypto.pbkdf2(
+                        derivedKey, salt, iterations,
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: derivedKey is the actual data encryption or decryption key used in the AES ctr cipher

File: lib/kms/common.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    /*
+     * cb(err, derivedKey, derivedIV)
+     */
+    static _deriveKey(cryptoScheme, dataKey, log, cb) {
+        if (cryptoScheme <= 1) {
+            /* we are not storing hashed human password.
+             * It's a random key, so 1 iteration and
+             * a fixed salt is enough for our usecase.
+             * don't change these without bumping
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: instead of 'these': the salt, the iteration number or the digest algorithm (sha1 here) without bumping the scheme number saved in the object metadata along with the dataKey.

File: lib/kms/common.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    /* block size is 128bit for AES-256-CTR */
+    static _aesBlockSize() {
+        return 16;
+    }
+
+    static createDataKey() {
+        return new Buffer(crypto.randomBytes(this._keySize()));
+    }
+
+    static _incrementIV(derivedIV, counter) {
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: please add unit tests for this function

File: lib/kms/common.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                    return;
+                }
+                const aesBlockSize = this._aesBlockSize();
+                const blocks = Math.floor(offset / aesBlockSize);
+                const toSkip = offset % aesBlockSize;
+                const iv = this._incrementIV(derivedIV, blocks);
+                const cipher = crypto.createDecipheriv(this._algorithm(),
+                                                       derivedKey, iv);
+                if (toSkip) {
+                    /* seek within the block to finish cipher initialization */
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: Above, we advanced to the latest boundary not greater than the offset amount.  Here we advance by the toSkip amount if necessary.

File: lib/kms/common.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                    cipher.write(dummyBuffer);
+                    cipher.read();
+                }
+                cb(null, cipher);
+            });
+    }
+
+    /*
+     * cb(err, cipher: ReadWritable.stream)
+     */
```
Comment (LaurenSpiegel) [inline lib/kms/common.js]: Included to give us flexibility for different functions for cipher and decipher.

File: lib/kms/file/backend.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+     */
+
+    /*
+     * cb(err, masterKeyId: string)
+     */
```
Comment (LaurenSpiegel) [inline lib/kms/file/backend.js]: Add comment: 
Using createDataKey here for purposes of createBucketKeyMem so that we do not need a separate function.

or just use random bytes.

File: lib/kms/file/backend.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            const newKey = common.createDataKey().toString('hex');
+            cb(null, newKey);
+        });
+    },
+
+    /*
+     * cb(err, cipheredDataKey: Buffer)
+     */
+    cipherDataKey: function cipherDataKeyMem(cryptoScheme,
+                                             masterKeyId,
```
Comment (LaurenSpiegel) [inline lib/kms/file/backend.js]: masterKeyId -- stored with the bucket metadata.  for file backend, the masterKeyId is the actual bucket key rather than the key to retrieve the actual key from a dictionary

File: lib/kms/file/backend.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+            common.createCipher(
+                cryptoScheme, masterKey, 0, log,
+                (err, cipher) => {
+                    if (err) {
+                        cb(err);
+                        return;
+                    }
+                    let cipheredDataKey =
+                            cipher.update(plainTextDataKey);
+                    const final = cipher.final();
```
Comment (LaurenSpiegel) [inline lib/kms/file/backend.js]: Here to ensure that any bytes remaining in the output of the stream are captured

File: lib/kms/in_memory/backend.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    /*
+     * Target implementation will be async in some aspects. let mimic it
+     */
+
+    /*
+     * cb(err, masterKeyId: string)
+     */
+    createBucketKey: function createBucketKeyMem(bucketName, log, cb) {
+        process.nextTick(() => {
+            kms[count] = common.createDataKey();
```
Comment (LaurenSpiegel) [inline lib/kms/in_memory/backend.js]: either comment that using the object data key for convenience or just create random bytes

File: lib/kms/wrapper.js [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+    client = file;
+    implName = 'fileKms';
+} else if (config.backends.kms === 'scality') {
+    throw new Error('KMS backend not implemented');
+} else {
+    throw new Error('KMS backend is not configured');
+}
+
+
+export default class kms {
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: KMS

File: lib/kms/wrapper.js [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+        const sseMasterKeyId =
+                  headers['x-amz-scal-server-side-encryption-aws-kms-key-id'];
+        /*
+         * the purpose of bucket level encryption is to exempt client
+         * from sending appropriate headers to trigger encryption.
+         * Since server side encryption with customer provided key needs
+         * the key for every access to the ciphered objects and we don't
+         * want to store it in the bucket metadata. It is not feasible
+         * in bucket level encryption.
+         */
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: The purpose of bucket level encryption is so that the client does not have to send appropriate headers to trigger encryption on each object put in an "encrypted bucket".
Customer provided keys are not feasible in this system because we do not want to store this key in the bucket metadata.

File: lib/kms/wrapper.js [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+                    return cb(err);
+                }
+                log.trace('cipher created by the kms');
+                return cb(null, cipher);
+            });
+    }
+
+    /*
+     * cb(err, decipher: ReadWritable.stream)
+     */
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: maybe call common.createDecipher directly from the code without this extra function layer

File: lib/kms/wrapper.js [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+            cb(null, serverSideEncryptionInfo);
+        } else {
+            /*
+             * no encryption
+             */
+            cb(null, null);
+        }
+    }
+
+    static createDataKey(log) {
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: is this supposed to mimic an async function?  It's sort of odd to have this function rather than just calling common.createDataKey() directly in createCipherBundle function below

File: lib/kms/wrapper.js [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+                }
+                log.trace('decipher created by the kms');
+                return cb(null, decipher);
+            });
+    }
+
+    /*
+     * cb(err, cipherBundle)
+     */
+    static createCipherBundle(serverSideEncryption,
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: I think it would be clearer to call this serverSideEncryptionInfo as you did above.

File: lib/kms/wrapper.js [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+            function cipherDataKey(next) {
+                return kms._cipherDataKey(cipherBundle.cryptoScheme,
+                                          serverSideEncryption.masterKeyId,
+                                          dataKey, log, next);
+            },
+            function createCipher(cipheredDataKey, next) {
+                cipherBundle.cipheredDataKey =
+                    cipheredDataKey.toString('base64');
+                return kms._createCipher(cipherBundle.cryptoScheme,
+                                         dataKey, 0, log, next);
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: should this be cipheredDataKey or dataKey?

File: lib/kms/wrapper.js [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+                log.error('error processing cipher bundle',
+                          { implName, error: err });
+            }
+            return cb(err, cipherBundle);
+        });
+    }
+    /*
+     * cb(err, decipherBundle)
+     */
+    static createDecipherBundle(serverSideEncryption, offset,
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: same as above -- serverSideEncryptionInfo

File: lib/kms/wrapper.js [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+                    serverSideEncryption.masterKeyId,
+                    serverSideEncryption.cipheredDataKey,
+                    log, next);
+            },
+            function createDecipher(plainTextDataKey, next) {
+                return kms._createDecipher(decipherBundle.cryptoScheme,
+                                           plainTextDataKey,
+                                           offset, log, next);
+            },
+            function finishDecipherBundle(decipher, next) {
```
Comment (LaurenSpiegel) [inline lib/kms/wrapper.js]: You could combine this with the final callback (or leave as is if you think clearer).

File: tests/unit/kms.js [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+                assert.strictEqual(err, null);
+                assert.strictEqual(sseInfo.cryptoScheme, 1);
+                assert.strictEqual(sseInfo.mandatory, true);
+                assert.strictEqual(sseInfo.algorithm, 'aws:kms');
+                assert.strictEqual(sseInfo.masterKeyId, masterKeyId);
+                done();
+            });
+    });
+
+    it('should not construct a sse info object on garbage', done => {
```
Comment (LaurenSpiegel) [inline tests/unit/kms.js]: 'should not construct a sse info object if x-amz-scal-server-side-encryption header contains invalid algorithm option'

File: tests/unit/kms.js [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+        kms.bucketLevelEncryption(
+            'dummyBucket', headers, log,
+            (err, sseInfo) => {
+                assert.strictEqual(err, null);
+                assert.strictEqual(sseInfo, null);
+                done();
+            });
+    });
+
+    it('should not construct a sse info object on no header', done => {
```
Comment (LaurenSpiegel) [inline tests/unit/kms.js]: 'should not construct a sse info object if no x-amz-scal-server-side-encryption header included with request'

File: tests/unit/kms.js [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    it('should cipher and decipher a datastream', done => {
+        _utestCreateBundlePair(log, (err, cipherBundle, decipherBundle) => {
+            assert.strictEqual(err, null);
+            cipherBundle.cipher.pipe(decipherBundle.decipher);
+            const target = new Buffer(10000);
+            target.fill('e');
+            cipherBundle.cipher.write(target);
+            const result = decipherBundle.decipher.read();
```
Comment (LaurenSpiegel) [inline tests/unit/kms.js]: I believe this will just read what is contained in the readable stream buffer.  So, you are relying on the high water mark default size of 16 KB for the test to pass.  You should note that or put a listener on the stream and concatenate the result.

**Comment by LaurenSpiegel** [comment]
@ironman-machine cool with this?

**Comment by LaurenSpiegel** [comment]
@ironman-machine let's start fresh

File: lib/kms/common.js [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+             * metadata along with the dataKey.
+             */
+            const salt = new Buffer('ItsTasty');
+            const iterations = 1;
+            crypto.pbkdf2(
+                dataKey, salt, iterations,
+                this._keySize(), 'sha1', (err, derivedKey) => {
+                    if (err) {
+                        log.error('pbkdf2 function failed on key derivation',
+                                  { err });
```
Comment (rahulreddy) [inline lib/kms/common.js]: Use error as key for err

File: lib/kms/common.js [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+                        cb(errors.InternalError);
```
Comment (rahulreddy) [inline lib/kms/common.js]: return callbacks instead if returning undefined.

File: lib/kms/common.js [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+                            if (err) {
+                                log.error(
+                                    'pbkdf2 function failed on IV derivation',
+                                    { err });
+                                cb(errors.InternalError);
+                                return;
+                            }
+                            // derivedKey is the actual data encryption or
+                            // decryption key used in the AES ctr cipher
+                            cb(null, derivedKey, derivedIV);
```
Comment (rahulreddy) [inline lib/kms/common.js]: return callback

File: lib/kms/common.js [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+                        });
+                });
+        } else {
+            log.error('Unknown cryptographic scheme', { cryptoScheme });
+            cb(errors.InternalError);
```
Comment (rahulreddy) [inline lib/kms/common.js]: Return callback here as well.

File: lib/kms/wrapper.js [hunk 32]
```diff
@@ -0,0 +1,255 @@
...
+                algorithm: sseAlgorithm,
+                masterKeyId: sseMasterKeyId,
+                mandatory: true,
+            };
+            cb(null, serverSideEncryptionInfo);
+        } else {
+            /*
+             * no encryption
+             */
+            cb(null, null);
```
Comment (rahulreddy) [inline lib/kms/wrapper.js]: If you are just checking for falsey values `return cb()` should be sufficient.

File: lib/kms/file/backend.js [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+                    }
+                    cb(null, cipheredDataKey);
+                });
+        });
+    },
+
+     /**
+      *
+      * @param {number} cryptoScheme - crypto scheme version number
+      * @param {string} masterKeyId - key to retrieve master key
```
Comment (alexandre-merle) [inline lib/kms/file/backend.js]: Confusing, we named it masterKeyId, but the key is encrypted and use to retrieve the master key

**Comment by alexandre-merle** [comment]
Minor stuff, but really good job :) :+1: (i will update the PR following the remaining comments)

File: lib/kms/common.js [hunk 34]
```diff
@@ -0,0 +1,165 @@
+import crypto from 'crypto';
+import { errors } from 'arsenal';
+
+/** Class exposing common createDataKey,
+createDecipher and createCipher functions. */
+export default class Common {
```
Comment (michael-zapata) [inline lib/kms/common.js]: If everything is static, why do a class instead of an object?

File: lib/kms/file/backend.js [hunk 35]
```diff
@@ -0,0 +1,109 @@
...
+      * @param {string} masterKeyId - master key; for the file backend
+      * the master key is the actual bucket master key rather than the key to
+      * retrieve the actual key from a dictionary
+      * @param {buffer} plainTextDataKey - data key
+      * @param {object} log - logger object
+      * @param {function} cb - callback
+      * @returns {undefined}
+      * @callback called with (err, cipheredDataKey: Buffer)
+      */
+    cipherDataKey: function cipherDataKeyMem(cryptoScheme,
```
Comment (michael-zapata) [inline lib/kms/file/backend.js]: decipher is identical aside from a single call, could we just parametrize the call?

Comment (michael-zapata) [inline lib/kms/file/backend.js]: Or well, the callback to that different function is similar, you can pretty much extract it.

File: lib/kms/common.js [hunk 36]
```diff
@@ -0,0 +1,165 @@
...
+     * @param {number} counter - quotient of the offset and blocksize
+     * @return {buffer} - the incremented IV
+     */
+    static _incrementIV(derivedIV, counter) {
+        const newIV = derivedIV;
+        const len = derivedIV.length;
+        let i = len - 1;
+        let ctr = counter;
+        while (ctr !== 0) {
+            const mod = (ctr + newIV[i]) % 256;
```
Comment (MathieuCassagne) [inline lib/kms/common.js]: I think the variable "mod' is not updated at each iteration, it's a const variable.
```es6
➜  ~ cat test.js 
var i = 0;
while (i < 10) {
	const a = i;
	++i;
	console.log(a);
}
➜  ~ node test.js 
0
0
0
0
0
0
0
0
0
0
```
```es6
➜  ~ cat test.js 
var i = 0;
while (i < 10) {
	var a = i;
	++i;
	console.log(a);
}
➜  ~ node test.js
0
1
2
3
4
5
6
7
8
9
```

File: lib/kms/in_memory/backend.js [hunk 37]
```diff
@@ -0,0 +1,104 @@
...
+      */
+    cipherDataKey: function cipherDataKeyMem(cryptoScheme,
+                                             masterKeyId,
+                                             plainTextDataKey,
+                                             log,
+                                             cb) {
+        process.nextTick(() => {
+            Common.createCipher(
+                cryptoScheme, kms[masterKeyId], 0, log,
+                (err, cipher) => {
```
Comment (michael-zapata) [inline lib/kms/in_memory/backend.js]: Still the same callback.

File: lib/kms/wrapper.js [hunk 38]
```diff
@@ -0,0 +1,255 @@
...
+            !serverSideEncryptionInfo.cryptoScheme) {
+            log.error('Invalid cryptographic information', { implName });
+            return cb(errors.InternalError);
+        }
+        const decipherBundle = {
+            cryptoScheme: serverSideEncryptionInfo.cryptoScheme,
+            decipher: null,
+        };
+        return async.waterfall([
+            function decipherDataKey(next) {
```
Comment (michael-zapata) [inline lib/kms/wrapper.js]: This function and the following ones are using parameters that are out of their own scope, it makes it hard to follow, and means we cannot rely on the function signature.

File: lib/kms/common.js [hunk 39]
```diff
@@ -0,0 +1,165 @@
...
+      * @returns {undefined}
+      * @callback called with (err, decipher: ReadWritable.stream)
+      */
+    static createDecipher(cryptoScheme, dataKey, offset, log, cb) {
+        this._deriveKey(
+            cryptoScheme, dataKey, log,
+            (err, derivedKey, derivedIV) => {
+                if (err) {
+                    log.warn('key derivation failed', { error: err });
+                    cb(err);
```
Comment (MathieuCassagne) [inline lib/kms/common.js]: Why not ```return cb(err);``` ?

File: lib/kms/common.js (hunk 36 above)
Comment (alexandre-merle) [inline lib/kms/common.js]: but for each iteration you can create the variable :) anyway you first example seem a little bit strange ^^

Comment (alexandre-merle) [inline lib/kms/common.js]: try to use 'use strict' and 'let i = 0' :)

File: lib/kms/common.js (hunk 34 above)
Comment (alexandre-merle) [inline lib/kms/common.js]: both are okay :)

File: lib/kms/file/backend.js (hunk 35 above)
Comment (alexandre-merle) [inline lib/kms/file/backend.js]: i don't get it :) what do you mean ? :)

File: lib/kms/common.js (hunk 36 above)
Comment (MathieuCassagne) [inline lib/kms/common.js]: Oh yes you're right, by using "use strict" there is a new variable at each iteration, my bad :)

Comment (alexandre-merle) [inline lib/kms/common.js]: :)

**Comment by MathieuCassagne** [comment]
:+1:

**Comment by alexandre-merle** [comment]
@michael-zapata  The modification for code duplication will be done in the final code, because this PR is actually the base PR for three others ones

**Comment by michael-zapata** [comment]
:+1: then.

**Comment by alexandre-merle** [comment]
Thank's :) Merging on behalf of @LaurenSpiegel

**Comment by alexandre-merle** [comment]
Squashing first :)
